### PR TITLE
@W-24102141: Support Refresh Token Rotation (RTR) with concurrency-safe refresh

### PR DESCRIPTION
**Pull request by iowillhoit**
### What does this PR do?

Adds Refresh Token Rotation (RTR) support to the refresh-token auth flow, and makes concurrent refreshes of the same auth safe:

- Persist the server-returned (rotated) refresh token instead of discarding it.
- Single-flight concurrent refreshes behind a dedicated `<authfile>.token-rotation.lock` so contenders adopt the rotated token instead of double-rotating (which RTR rejects with "Token request is already being processed").
- Tighten JWT vs refresh-token routing and clear a stale `privateKey` when a JWT auth is later re-authed via web.

Design rationale and edge-case handling live in the inline comments, so this body stays short.

## Testing

**Unit** (no org needed):

```
yarn mocha test/unit/org/authInfo.test.ts
```

**Concurrency NUT** (hits a real org, rotates real tokens, OFF by default). Requires a **web/refresh-token-authed** org (not JWT) whose connected app has **RTR enabled**:

```
# Reuse an existing web-authed scratch org (alias or username):
RTR_NUT=1 RTR_SCRATCH_USERNAME=<your-web-authed-org> \
  yarn mocha test/nut/refreshTokenRotation.nut.ts --timeout 600000

# Or create + clean up a throwaway one from a WEB-AUTHED dev hub:
RTR_NUT=1 RTR_DEVHUB=<your-web-authed-hub> \
  yarn mocha test/nut/refreshTokenRotation.nut.ts --timeout 600000
```

Fires 30 concurrent refreshes (override with `RTR_CONCURRENCY`) and asserts: all succeed, no contention errors, exactly one (few) real rotation with the rest adopting, token actually rotated, and no lockout afterward.

### What issues does this PR fix or reference?
[@W-24102141@](https://gus.my.salesforce.com/apex/ADM_WorkLocator?bugorworknumber=W-24102141)

### CODE REVIEW

File: messages/core.md [hunk 1]
```diff
@@ -34,6 +34,14 @@ Errors encountered:
 
 Error authenticating with the refresh token due to: %s
 
+# refreshTokenAuthError.actions
+
+- The stored refresh token is no longer valid (it may have expired, been revoked, or been rotated). Re-authenticate to the org, then retry. For example, run "sf org login web --alias <your-alias>"
+
+# refreshTokenRotationTimeoutError
+
+Timed out waiting for another process to finish rotating the refresh token for %s. Another process held the rotation lock for the entire wait without completing. Retry the command. If it persists, another "sf" process is likely stuck mid-refresh: stop any other running "sf" processes, then retry.
```
Comment (WillieRuemmele) [inline messages/core.md]: I think we should be careful about using `sf` commands as examples, because `sfdx-core` is used by other consumers and showing these error messages could be confusing, or counter-intuitive?

File: src/org/authInfo.ts [hunk 2]
```diff
@@ -1035,10 +1039,22 @@ export class AuthInfo extends AsyncOptionalCreatable<AuthInfo.Options> {
...
           authConfig = await this.authJwt(options);
         } else if (!options.authCode && options.refreshToken) {
-          // refresh token flow (from sfdxUrl or OAuth refreshFn)
+          // refresh token flow (from sfdxUrl or OAuth refreshFn).
+          // RTR CAUTION: this POST rotates the token server-side and invalidates the old one immediately.
+          // Everything from here until the caller's save() (determineIfDevHub, orgs.read, update/encrypt,
+          // determineOrg) runs on borrowed time -- a throw in that window strands the rotated token on the
+          // wire while the old one is already dead server-side, permanently breaking the auth until re-login.
+          // determineIfDevHub and determineOrg swallow their own errors, so the practical window is small,
+          // but it is non-zero. A save-early seam for this branch would close it (tracked as a follow-up).
```
Comment (WillieRuemmele) [inline src/org/authInfo.ts]: any reason to not do this now?
> (tracked as a follow-up)

File: messages/core.md (hunk 1 above)
Comment (iowillhoit) [reply messages/core.md]: Good call, updating

File: src/org/authInfo.ts (hunk 2 above)
Comment (iowillhoit) [reply src/org/authInfo.ts]: It appears to be low risk. Most of the errors on that path don't throw (swallowed). 
It also moves the save location to a place that affects non-RTR flows, so the "fix" might be riskier than than the edgecase.
I'll take another look at this tomorrow just to confirm

Comment (iowillhoit) [reply src/org/authInfo.ts]: Ah, this is simpler than i thought, I'll make this change and push it shortly @WillieRuemmele

File: src/stateAggregator/accessors/orgAccessor.ts [hunk 3]
```diff
@@ -157,6 +176,16 @@ export abstract class BaseOrgAccessor<T extends ConfigFile, P extends ConfigCont
     return config ? config.stat() : null;
   }
 
+  /**
+   * Return the absolute path to the auth file for a given username. Does not require the file to have been
+   * read (resolves purely from the username and the global state directory).
+   *
+   * @param username
+   */
+  public getPath(username: string): string {
```
Comment (mshanemc) [inline src/stateAggregator/accessors/orgAccessor.ts:185]: why not just have one, and let it be public?

**Comment by WillieRuemmele** [comment]
## QA Results — RTR (Refresh Token Rotation) ✅ 

Tested against a web-authed devhub (`admin@integrationtesthubna40.org`) with a fresh 1-day scratch org.

### NUT (in-process, 30 concurrent refreshes)
```
RTR_NUT=1 RTR_DEVHUB=admin@integrationtesthubna40.org yarn mocha test/nut/refreshTokenRotation.nut.ts --timeout 600000
```
- **30 concurrent connections, 1 real refresh call** — lock serialization working as expected
- Refresh token rotated, all assertions passed

### CLI smoke test

| Scenario | Result |
|----------|--------|
| Scratch org created from web-authed devhub | Refresh token present, no privateKey |
| Single command with forced refresh | Token rotated |
| 5 concurrent CLI commands with forced refresh | All 5 succeeded, 0 failures, token rotated once |
| Stale lock files after concurrent run | None |
| Corrupted access token recovery | Recovered and token rotated |

### What was tested
- Invalidated the on-disk access token before each scenario to force a real refresh + RTR rotation
- Verified the refresh token value changed in the auth file (`~/.sfdx/<username>.json`) after each rotation
- Confirmed no `.token-rotation.lock` files lingered after concurrent runs
- Confirmed JWT-authed orgs are unaffected (no refresh token = RTR code path not entered)

<details>
<summary>QA smoke test script (rtr-qa.sh)</summary>

```bash
#!/bin/bash
# RTR QA smoke test — run from any directory with a web-authed devhub.
# Usage: RTR_DEVHUB=<your-web-authed-hub> bash rtr-qa.sh

set -euo pipefail

DEVHUB="${RTR_DEVHUB:?Set RTR_DEVHUB to a web-authed devhub username/alias}"
ALIAS="rtr-qa-$$"
SCRATCH_CREATED=false

# Create a temp SFDX project so sf org create scratch works
PROJECT_DIR=$(mktemp -d -t rtr-qa-XXXX)
mkdir -p "$PROJECT_DIR/force-app" "$PROJECT_DIR/config"
echo '{"packageDirectories":[{"path":"force-app","default":true}],"namespace":""}' > "$PROJECT_DIR/sfdx-project.json"
echo '{"orgName":"RTR QA","edition":"Developer"}' > "$PROJECT_DIR/config/project-scratch-def.json"

cleanup() {
  if $SCRATCH_CREATED; then
    echo "--- Cleaning up scratch org $ALIAS ---"
    sf org delete scratch -o "$ALIAS" --no-prompt 2>/dev/null || true
  fi
  rm -rf "$PROJECT_DIR"
}
trap cleanup EXIT

invalidate_access_token() {
  jq '.accessToken = "00Dxx0000000000!INVALID_forcing_refresh"' "$AUTH_FILE" > "${AUTH_FILE}.tmp" \
    && mv "${AUTH_FILE}.tmp" "$AUTH_FILE"
}

echo "=== 1. Create scratch org from web-authed devhub ==="
(cd "$PROJECT_DIR" && sf org create scratch --definition-file config/project-scratch-def.json --target-dev-hub "$DEVHUB" --duration-days 1 --wait 10 --alias "$ALIAS" --json) | jq '{status, username: .result.username}'
SCRATCH_CREATED=true

echo ""
echo "=== 2. Confirm org is web/refresh-token authed (not JWT) ==="
USERNAME=$(sf org display -o "$ALIAS" --json | jq -r '.result.username')
AUTH_FILE=$(find ~/.sfdx ~/.sf -name "${USERNAME}.json" 2>/dev/null | head -1)
if [ -z "$AUTH_FILE" ]; then
  echo "ERROR: Could not find auth file for $USERNAME"
  exit 1
fi
echo "Auth file: $AUTH_FILE"
jq '{refreshToken: (if .refreshToken then "present" else "MISSING" end), privateKey: (if .privateKey then "present" else "NONE" end)}' "$AUTH_FILE"

echo ""
echo "=== 3. Record initial refresh token ==="
BEFORE=$(jq -r '.refreshToken // empty' "$AUTH_FILE")
if [ -z "$BEFORE" ]; then
  echo "ERROR: No refresh token found — org may be JWT-authed. RTR cannot be tested."
  exit 1
fi
echo "Initial refreshToken (last 10 chars): ...${BEFORE: -10}"

echo ""
echo "=== 4. Single command with forced refresh — does token rotate? ==="
invalidate_access_token
echo "Invalidated access token to force refresh."
sf data query -o "$ALIAS" -q "SELECT Id FROM Organization" --json | jq '{status, orgId: .result.records[0].Id}'
AFTER=$(jq -r '.refreshToken' "$AUTH_FILE")
echo "Before (last 10): ...${BEFORE: -10}"
echo "After  (last 10): ...${AFTER: -10}"
if [ "$BEFORE" != "$AFTER" ]; then
  echo "RESULT: Token ROTATED"
else
  echo "RESULT: Token did NOT rotate (RTR may not be enabled on the connected app)"
fi

echo ""
echo "=== 5. Concurrent commands with forced refresh (5 parallel queries) ==="
invalidate_access_token
echo "Invalidated access token to force refresh."
BEFORE_CONCURRENT=$(jq -r '.refreshToken' "$AUTH_FILE")
PIDS=()
RESULTS_DIR=$(mktemp -d)
for i in {1..5}; do
  (
    sf data query -o "$ALIAS" -q "SELECT Id FROM Organization" --json > "$RESULTS_DIR/$i.json" 2>&1
    echo $? > "$RESULTS_DIR/$i.exit"
  ) &
  PIDS+=($!)
done

for pid in "${PIDS[@]}"; do
  wait "$pid" || true
done

FAIL_COUNT=0
echo "Concurrent results:"
for i in {1..5}; do
  EXIT_CODE=$(cat "$RESULTS_DIR/$i.exit" 2>/dev/null || echo "?")
  STATUS=$(jq -r '.status // .name // "parse-error"' "$RESULTS_DIR/$i.json" 2>/dev/null || echo "error")
  echo "  job $i: exit=$EXIT_CODE status=$STATUS"
  [ "$EXIT_CODE" != "0" ] && FAIL_COUNT=$((FAIL_COUNT + 1))
done
echo "Failures: $FAIL_COUNT / 5"

AFTER_CONCURRENT=$(jq -r '.refreshToken' "$AUTH_FILE")
echo "Before (last 10): ...${BEFORE_CONCURRENT: -10}"
echo "After  (last 10): ...${AFTER_CONCURRENT: -10}"
if [ "$BEFORE_CONCURRENT" != "$AFTER_CONCURRENT" ]; then
  echo "RESULT: Token ROTATED during concurrent test"
else
  echo "RESULT: Token did NOT rotate during concurrent test"
fi
rm -rf "$RESULTS_DIR"

echo ""
echo "=== 6. Check for stale lock files ==="
LOCK_PATH="${AUTH_FILE}.token-rotation.lock"
if [ -e "$LOCK_PATH" ]; then
  echo "WARNING: Lock file exists at $LOCK_PATH"
else
  echo "OK: No lingering lock file"
fi

echo ""
echo "=== 7. Corrupt access token again, verify recovery ==="
BEFORE_RECOVERY=$(jq -r '.refreshToken' "$AUTH_FILE")
invalidate_access_token
echo "Corrupted access token, running query..."
if sf data query -o "$ALIAS" -q "SELECT Id FROM Organization" --json | jq '{status}'; then
  AFTER_RECOVERY=$(jq -r '.refreshToken' "$AUTH_FILE")
  if [ "$BEFORE_RECOVERY" != "$AFTER_RECOVERY" ]; then
    echo "RESULT: Recovered and token ROTATED"
  else
    echo "RESULT: Recovered but token did not rotate"
  fi
else
  echo "RESULT: FAILED to recover from corrupted access token"
fi

echo ""
echo "=== Done ==="
```

</details>

File: src/stateAggregator/accessors/orgAccessor.ts (hunk 3 above)
Comment (iowillhoit) [reply src/stateAggregator/accessors/orgAccessor.ts:185]: I could make `parseFilename` public, but I think it's a bad name for a public method. It's not "parsing a filename", it's taking a username and returning an absolute path

Comment (mshanemc) [reply src/stateAggregator/accessors/orgAccessor.ts:185]: I was thinking, get rid of parseFilename and have getPath be the only

**Review by mshanemc — APPROVED** [review]

**Review by WillieRuemmele — APPROVED** [review]
